goldmane/pkg/aggregator/stream_manager.go

```diff
@@ -0,0 +1,137 @@
+package aggregator
+
+import (
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/goldmane/pkg/internal/types"
+	"github.com/projectcalico/calico/goldmane/proto"
+)
+
+type Stream struct {
+	id    string
+	flows chan *proto.Flow
+	done  chan<- string
+	req   streamRequest
+
+	bucketsToAggregate int
+	rolloverCounter    int
+}
+
+func (s *Stream) Close() {
+	s.done <- s.id
+}
+
+func (s *Stream) Flows() <-chan *proto.Flow {
+	return s.flows
+}
+
+func (s *Stream) Send(f *proto.Flow) {
+	select {
+	case s.flows <- f:
+	case <-time.After(5 * time.Second):
+		logrus.WithField("id", s.id).Warn("Timed out sending flow to stream")
+	}
+}
+
+func (s *Stream) inc() {
+	s.rolloverCounter++
+}
+
+func (s *Stream) shouldEmit() bool {
+	return s.rolloverCounter%s.bucketsToAggregate == 0
+}
+
+func NewStreamManager() *streamManager {
+	maxStreams := 100
+	return &streamManager{
+		streams:         make(map[string]*Stream),
+		closedStreamsCh: make(chan string, maxStreams),
+		maxStreams:      maxStreams,
+	}
+}
+
+type streamManager struct {
+	// streams is a registry of active streams being served by the aggregator. Stream data
+	// is published to these streams on rollover.
+	streams map[string]*Stream
+
+	// maxStreams configured the maximum number of concurrent streams that can be active.
+	// If this limit is reached, new streams will be rejected.
+	maxStreams int
+
+	// closedStreamsCh is a channel on which to receive UUIDs of streams that have been closed.
+	closedStreamsCh chan string
+}
+
+func (m *streamManager) closedStreams() chan string {
+	return m.closedStreamsCh
+}
+
+func (m *streamManager) register(req streamRequest) *Stream {
+	if m.maxStreams > 0 && len(m.streams) >= m.maxStreams {
+		logrus.WithField("max", m.maxStreams).Warn("Max streams reached, rejecting new stream")
+		return nil
+	}
+
+	stream := &Stream{
+		id:                 uuid.NewString(),
+		flows:              make(chan *proto.Flow, 100),
+		done:               m.closedStreamsCh,
+		req:                req,
+		bucketsToAggregate: 1,
+	}
+	m.streams[stream.id] = stream
+
+	logrus.WithField("id", stream.id).Debug("Registered new stream")
+	return stream
+}
+
+func (m *streamManager) close(id string) {
+	s, ok := m.streams[id]
+	if !ok {
+		logrus.WithField("id", id).Warn("Asked to close unknown stream")
+		return
+	}
+	logrus.WithField("id", id).Debug("Closing stream")
+	close(s.flows)
+	delete(m.streams, id)
+}
+
+func (m *streamManager) rollover(ring *BucketRing, diachronics map[types.FlowKey]*types.DiachronicFlow) {
+	// Each stream may have requested aggregation across a different number of buckets, so the individual stream
+	// objects are responsible for tracking their own state - whether or not it is ready to emit data, and how many buckets
+	// back it wants to go.
+	logrus.Debug("Checking if any stream should emit data")
```

Comment:
What would you think about keeping the logic with the Ring, instead of requiring this coupling between the types?

You could just send the Diachronics on a channel to emit and let the ring figure out what ones it's sending.

We might not need to care about tracking the rollovers between streams, and just keep the iteration constant for all streams. For instance, if we're iterating on 6 buckets, and a stream is created on the 5th bucket, I think it would be fine to send back the 6 buckets worth of data even though it includes flows that were created before the stream started.

At least we could start with this simplish / consistent implementation and get a little more complex if we realise this isn't enough?

Reply:
> For instance, if we're iterating on 6 buckets, and a stream is created on the 5th bucket, I think it would be fine to send back the 6 buckets worth of data even though it includes flows that were created before the stream started.

The problem is that, in theory, two streams need not have the same aggregation window. 

User 1 logs into the UI, says "Show me flows ever 15s" and user 2 logs in and says "Show me flows every minute" - those result in two separate Stream objects, each sending flows at a different rate and aggregating across different sets of buckets.